repository/src/main/java/com/epam/esm/repository/impl/CertificateRepositoryImpl.java

```diff
@@ -3,117 +3,153 @@
+            Predicate predicate = criteriaBuilder.like(certificateRoot.get(DESCRIPTION), partialName);
+            predicateList.add(predicate);
+        }
+
+        //Condition on which parameter to sort by
+        if (StringUtils.isNotEmpty(searchCriteria.getSortByParameter())) {
+            String partialName = String.format(PARTIAL_STRING, searchCriteria.getSortByParameter());
+            Predicate predicate = criteriaBuilder.like(certificateRoot.get(PARAMETER), partialName);
+            predicateList.add(predicate);
+        }
+
+        //Condition by order type
+        if (searchCriteria.getOrderType() != null) {
+            Order order = switch (searchCriteria.getOrderType()) {
+                case ASC -> criteriaBuilder.asc(certificateRoot.get(PARAMETER));
+                case DESC -> criteriaBuilder.desc(certificateRoot.get(PARAMETER));
+            };
+            query.orderBy(order);
+        }
+
+        query = query.select(certificateRoot).distinct(true);
+
+        //Condition to find number of tags if exists
+        if (searchCriteria.getTagList() != null) {
+            Predicate inListPredicate = join.get(NAME).in(searchCriteria.getTagList());
+            predicateList.add(inListPredicate);
+
+            query = query.where(predicateList.toArray(new Predicate[0]))
+                    .groupBy(
+                            certificateRoot.get(ID),
+                            certificateRoot.get(NAME),
+                            certificateRoot.get(DESCRIPTION),
+                            certificateRoot.get(PRICE),
+                            certificateRoot.get(DURATION),
+                            certificateRoot.get(CREATE_DATE),
+                            certificateRoot.get(LAST_UPDATE_DATE)
+                    ).having(
+                            criteriaBuilder.equal(
+                                    criteriaBuilder.countDistinct(join.get(ID)),
+                                    searchCriteria.getTagList().size()
+                            )
+                    );
+        } else {
+            query = query.where(predicateList.toArray(new Predicate[0]));
+        }
 
-        QueryBuilder sqlClause = new QueryBuilder.Builder(searchCriteria, SELECT_ALL_CERTIFICATES)
-                .buildWhereClause(parameters)
-                .buildOrderByClause(searchCriteria)
-                .build();
-        return namedParameterJdbcTemplate.query(sqlClause.getNewString(), parameters, rowMapper);
+        return entityManager.createQuery(query).getResultList();
     }
 
     @Override
     public Optional<Certificate> findById(Long id) {
-        SqlParameterSource parameterSource = new MapSqlParameterSource().addValue(ID_PARAMETER, id);
-        List<Certificate> certificates = namedParameterJdbcTemplate.query(SELECT_CERTIFICATE_BY_ID, parameterSource, rowMapper);
-
-        return certificates.stream().findFirst();
+        return Optional.of(entityManager.find(Certificate.class, id));

```

Comment:
Read the difference between the of() and ofNullable() methods

Reply:
fixed